File: 2024/day1/node/main.ts
import fs from 'node:fs';

function panic(err: Error) {
    console.error(err);
    process.exit(1);
}

async function getInput(): Promise<string> {
    try {
        const data = await fs.promises.readFile('input.txt', 'utf-8')
        return data
    } catch (error) {
        panic(error);
    }
    return ''; // Unreachable anyway, just to make the compiler happy
}

async function parseInput(input: string): Promise<[number[], number[]]> {
    const lines = input.split('\n')
    const left_ids: number[] = []
    const right_ids: number[] = []
    lines.forEach((line) => {
        if (line === '') {
            return
        }
        const parts = line.split('   ')
        left_ids.push(parseInt(parts[0]))
        right_ids.push(parseInt(parts[1]))
    })

    return [left_ids, right_ids]
}

async function part_one() {
    const input = await getInput();
    const [left_ids, right_ids] = await parseInput(input);
    left_ids.sort();
    right_ids.sort();

    let total = 0;

    for (let i = 0; i < left_ids.length; i++) {
        total += Math.abs(left_ids[i] - right_ids[i]);
    }

    console.log("Part one: ", total);
}

async function part_two() {
    const input = await getInput();
    const [left_ids, right_ids] = await parseInput(input);
    let right_map = new Map<number, number>();
    right_ids.forEach((id) => {
        let old = right_map.get(id) ?? 0;
        right_map.set(id, old + 1);
    })

    let total = 0;

    for (let i = 0; i < left_ids.length; i++) {
        let id = left_ids[i];
        let score = right_map.get(id) ?? 0;
        total += score * id
    }

    console.log("Part two: ", total);
}

async function main() {
    console.log('Node solution');
    await part_one();
    await part_two();
    console.log('------------------------------------------------------')
}

//Run
await main()

File: 2024/day1/go/main.go
package main

import (
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
	"strings"
)

func get_input() string {
	data, err := os.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}

	return string(data)
}

func parseInt(raw string) int {
	num, err := strconv.Atoi(raw)
	if err != nil {
		panic(err)
	}
	return num
}

func parse_input(input string) ([]int, []int) {
	lines := strings.Split(input, "\n")
	left_ids := make([]int, 0)
	right_ids := make([]int, 0)
	for _, line := range lines {
		if line == "" { // EOF
			break
		}
		parts := strings.Split(line, "   ")
		left_ids = append(left_ids, parseInt(parts[0]))
		right_ids = append(right_ids, parseInt(parts[1]))
	}
	return left_ids, right_ids
}
func part_one() {
	input := get_input()
	left_ids, right_ids := parse_input(input)
	sort.Ints(left_ids)
	sort.Ints(right_ids)
	total := 0
	for i := range left_ids {
		left_id := left_ids[i]
		right_id := right_ids[i]
		total += (int)(math.Abs(float64(left_id) - float64(right_id)))
	}

	fmt.Printf("Part one: %d\n", total)
}

func part_two() {
	input := get_input()
	left_ids, right_ids := parse_input(input)
	right_map := make(map[int]int)
	for _, id := range right_ids {
		if value, ok := right_map[id]; ok {
			right_map[id] = value + 1
		} else {
			right_map[id] = 1
		}
	}

	total := 0
	for _, id := range left_ids {
		if value, ok := right_map[id]; ok {
			total += value * id
		}
	}

	fmt.Printf("Part two: %d\n", total)
}

func main() {
	fmt.Println("Go solutions")
	part_one()
	part_two()
}
